File: gym_management/gym_management/report/fitness_journey_report/fitness_journey_report.py
import frappe


def execute(filters=None):
    filters = filters or {}

    columns = [
        {
            "label": "Date",
            "fieldname": "date",
            "fieldtype": "Date",
            "width": 120,
        },
        {
            "label": "Gym Member",
            "fieldname": "gym_member",
            "fieldtype": "Link",
            "options": "Gym Member",
            "width": 180,
        },
        {
            "label": "Weight (kg)",
            "fieldname": "weight",
            "fieldtype": "Float",
            "width": 120,
        },
        {
            "label": "Calories Burned",
            "fieldname": "calories",
            "fieldtype": "Int",
            "width": 150,
        },
        {
            "label": "Workout Plan",
            "fieldname": "workouts",
            "fieldtype": "Link",
            "options": "Gym Workout Plan",
            "width": 180,
        },
    ]

    query_filters = {}
    if filters.get("gym_member"):
        query_filters["gym_member"] = filters.get("gym_member")

    records = frappe.get_all(
        "Fitness Journey",
        fields=[
            "date",
            "gym_member",
            "weight",
            "calories",
            "workouts",
        ],
        filters=query_filters,
        order_by="date asc",
    )

    data = []
    labels = []
    weight_values = []
    calorie_values = []

    for row in records:
        data.append(row)

        label = row.get("date") or "N/A"
        labels.append(label)
        weight_values.append(row.get("weight") or 0)
        calorie_values.append(row.get("calories") or 0)

    # ✅ BAR CHART (WORKS FOR 1 OR MANY RECORDS)
    chart = {
        "data": {
            "labels": labels,
            "datasets": [
                {
                    "name": "Weight Progress",
                    "values": weight_values,
                },
                {
                    "name": "Calories Burned",
                    "values": calorie_values,
                },
            ],
        },
        "type": "bar",
    }

    return columns, data, None, chart

File: gym_management/gym_management/report/fitness_journey_report/fitness_journey_report.js
frappe.query_reports["Fitness Journey Report"] = {
    filters: [
        {
            fieldname: "gym_member",
            label: "Gym Member",
            fieldtype: "Link",
            options: "Gym Member",
        },
        {
            fieldname: "month",
            label: "Month",
            fieldtype: "Select",
            options: [
                "",
                "January",
                "February",
                "March",
                "April",
                "May",
                "June",
                "July",
                "August",
                "September",
                "October",
                "November",
                "December",
            ],
            default: "",
        },
        {
            fieldname: "year",
            label: "Year",
            fieldtype: "Int",
            default: new Date().getFullYear(),
        },
    ],

    formatter: function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname === "gym_member" && data && data.gym_member) {
            value = `
                <a href="#"
                   class="gym-member-link"
                   data-member="${data.gym_member}">
                   ${value}
                </a>`;
        }

        return value;
    },

    onload: function (report) {
        $(document)
            .off("click", ".gym-member-link")
            .on("click", ".gym-member-link", function (e) {
                e.preventDefault();

                const member = $(this).data("member");
                report.set_filter_value("gym_member", member);
                report.refresh();
            });
    },
};
